bundle/util.go

```diff
@@ -0,0 +1,396 @@
+//
+// Copyright (c) 2018 Red Hat, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package bundle
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"encoding/json"
+
+	schema "github.com/lestrrat/go-jsschema"
+)
+
+type formItem struct {
+	Key   string        `json:"key,omitempty"`
+	Title string        `json:"title,omitempty"`
+	Type  string        `json:"type,omitempty"`
+	Items []interface{} `json:"items,omitempty"`
+}
+
+// ConvertPlanToSchema
+func ConvertPlansToSchema(plans []Plan) ([]SchemaPlan, error) {
+	brokerPlans := make([]SchemaPlan, len(plans))
+	i := 0
+	for _, plan := range plans {
+		schemas, err := parametersToSchema(plan)
+		if err != nil {
+			return nil, err
+		}
+		brokerPlans[i] = SchemaPlan{
+			ID:          plan.ID,
+			Name:        plan.Name,
+			Description: plan.Description,
+			Metadata:    extractBrokerPlanMetadata(plan),
+			Free:        plan.Free,
+			Bindable:    plan.Bindable,
+			UpdatesTo:   plan.UpdatesTo,
+			Schemas:     schemas,
+		}
+		i++
+	}
+	return brokerPlans, nil
+}
+
+func planUpdatable(apbPlans []Plan) bool {
+	for _, plan := range apbPlans {
+		if len(plan.UpdatesTo) > 0 {
+			return true
+		}
+	}
+	return false
+}
+
+func extractBrokerPlanMetadata(apbPlan Plan) map[string]interface{} {
+	metadata, err := initMetadataCopy(apbPlan.Metadata)
+
+	if err != nil {
+		return apbPlan.Metadata
+	}
+
+	instanceFormDefn := createFormDefinition(apbPlan.Parameters)
+	bindingFormDefn := createFormDefinition(apbPlan.BindParameters)
+
+	metadata["schemas"] = map[string]interface{}{
```

Comment:
Funny to see how verbose this declaration is after working in python for a while.